### PR TITLE
[feat] Added Lusca middleware for CSRF [fixes #828]

### LINKED ISSUE #828: CSRF protection middleware for the app?
are there any plans to implement CSRF protection as the app uses cookie based authentication?

### PR DESCRIPTION
**Pull request by yilenpan**
Added Lusca middleware for CSRF

### CODE REVIEW

File: config/env/default.js [hunk 1]
```diff
@@ -26,6 +26,15 @@ module.exports = {
   // sessionKey is set to the generic sessionId key used by PHP applications
   // for obsecurity reasons
   sessionKey: 'sessionId',
+  // Lusca CSRF config
+  csrf: {
+    csrf: false,
+    csp: { /* ... */},
```
Comment (ilanbiala) [inline config/env/default.js]: Why the comment?

File: config/env/default.js [hunk 2]
```diff
@@ -26,6 +26,15 @@ module.exports = {
...
+    xframe: 'SAMEORIGIN',
+    p3p: 'ABCDEF',
+    hsts: {maxAge: 31536000, includeSubDomains: true, preload: true},
```
Comment (ilanbiala) [inline config/env/default.js]: What is that number? Put a comment to clarify.

File: package.json [hunk 3]
```diff
@@ -42,6 +42,7 @@
     "helmet": "~0.9.1",
     "jasmine-core": "^2.3.4",
     "lodash": "^3.10.0",
+    "lusca": "^1.3.0",
```
Comment (ilanbiala) [inline package.json]: ~

File: config/env/default.js [hunk 4]
```diff
@@ -26,6 +26,17 @@ module.exports = {
   // sessionKey is set to the generic sessionId key used by PHP applications
   // for obsecurity reasons
   sessionKey: 'sessionId',
+  // Lusca CSRF config
+  csrf: {
```
Comment (lirantal) [inline config/env/default.js]: maybe this should be named lusca config since it's more than just csrf?

File: config/env/default.js [hunk 5]
```diff
@@ -26,6 +26,17 @@ module.exports = {
...
+    csrf: false,
```
Comment (lirantal) [inline config/env/default.js]: Why is CSRF false by default?
Also, doesn't CSRF require updates on the frontend part to send a token?

Comment (codydaig) [inline config/env/default.js]: Not sure about the client side token, however, I like csrf default to false, that way it's a non breaking change and you'd have to enable it and configure it if you wanted it. 

@yilenpan Meet-up with me after-hours sometime and we can investigate the client side token further.

**Comment by pgrodrigues** [comment]
I didn't test but I believe the server route tests might break unless the csrf token is passed. Just a heads-up before merging this in.

File: config/env/default.js (hunk 5 above)
Comment (yilenpan) [inline config/env/default.js]: @codydaig sounds good

File: config/env/default.js [hunk 6]
```diff
@@ -26,6 +26,17 @@ module.exports = {
...
   sessionKey: 'sessionId',
+  // Lusca config
+  csrf: {
+    csrf: false,
+    csp: { /* Content Security Policy object */},
+    xframe: 'SAMEORIGIN',
+    p3p: 'ABCDEF',
+    hsts: { maxAge: 31536000, // Forces HTTPS for one year
+            includeSubDomains: true,
+            preload: true },
```
Comment (ilanbiala) [inline config/env/default.js]: newline

Comment (yilenpan) [inline config/env/default.js]: Between line 28/29?

Comment (ilanbiala) [inline config/env/default.js]: The comment is on 37, the newline should be before the `}` in 37.

File: config/env/default.js [hunk 7]
```diff
@@ -26,6 +26,19 @@ module.exports = {
   // sessionKey is set to the generic sessionId key used by PHP applications
   // for obsecurity reasons
   sessionKey: 'sessionId',
+  // Lusca config
+  csrf: {
```
Comment (lirantal) [inline config/env/default.js]: maybe this should be named lusca config since it's more than just csrf?

**Comment by lirantal** [comment]
@yilenpan see my comments.

Overall, this is a great addition.
We do need to augment it with a frontend PR to add those tokens for all those API calls that we're making, and then we can enable CSRF protection on too.

**Comment by codydaig** [comment]
Bumping milestone to 0.5.0 so we're not holding up 0.4.2

**Comment by ilanbiala** [comment]
@lirantal any progress on this?

**Comment by lirantal** [comment]
I haven't had a chance to try out the changes introduced in this PR yet.
This is till on my to-do list.

**Comment by ilanbiala** [comment]
@yilenpan can you rebase this? @lirantal can you review this one more time before we merge?

**Comment by lirantal** [comment]
@ilanbiala sure but let's see if @yilenpan wishes to get back to this for the required fixes (been a couple of months since), otherwise I'll take over.

**Comment by yilenpan** [comment]
Hey fellas, sorry for being a little MIA, I've rebased pushed up the new changes. Seems to trip up the linter for some reason, so I'll try and figure that one out too :(

**Comment by yilenpan** [comment]
Hmm.. Looks like there's an error in the `changeProfilePicture` method in `/modules/users/server/controllers/users/users.profile.server.controller.js`, could be the config?

File: config/env/default.js (hunk 7 above)
Comment (lirantal) [inline config/env/default.js]: @yilenpan ?

**Comment by yilenpan** [comment]
Hey guys, middleware has been added and build is passing. 

Best

-Y

File: package.json [hunk 8]
```diff
@@ -43,6 +43,7 @@
     "helmet": "~0.9.1",
     "jasmine-core": "~2.3.4",
     "lodash": "~3.10.0",
+    "lusca": "^1.3.0",
```
Comment (codydaig) [inline package.json]: `~`

**Comment by codydaig** [comment]
Left a line comment. Also, make sure the commit message matches the guidelines outlined in CONTRIBUTING.md.

**Comment by yilenpan** [comment]
The Node.js 0.12 macosx build failed when trying to install protractor, all other builds passed. May want to rerun that test?

**Comment by ilanbiala** [comment]
Reran it.

**Comment by ilanbiala** [comment]
@lirantal LGTM, wanna take a look?

**Comment by lirantal** [comment]
Great.
Some follow-up items for afterwards:
1. Frontend side needs to make use of the CSRF tokens
2. @yilenpan didn't change the config object from csrf to lusca and I think we should but I'm in favor of merging often, and don't want to delay this PR due to my personal favorites.

**Comment by ilanbiala** [comment]
Who is going to implement the frontend for the tokens?

**Pull request merged**